logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/auth/sso/SSOInitiateLoginUseCase.kt

```diff
@@ -0,0 +1,50 @@
+package com.wire.kalium.logic.feature.auth.sso
+
+import com.wire.kalium.logic.CoreFailure
+import com.wire.kalium.logic.NetworkFailure
+import com.wire.kalium.logic.configuration.ServerConfig
+import com.wire.kalium.logic.data.auth.login.SSOLoginRepository
+import com.wire.kalium.logic.feature.auth.ValidateUUIDUseCase
+import com.wire.kalium.logic.functional.suspending
+import com.wire.kalium.network.exceptions.KaliumException
+import io.ktor.http.HttpStatusCode
+
+sealed class SSOInitiateLoginResult {
+    data class Success(val requestUrl: String) : SSOInitiateLoginResult()
+
+    sealed class Failure : SSOInitiateLoginResult() {
+        object InvalidCode : Failure()
+        object InvalidRedirect : Failure()
+        class Generic(val genericFailure: CoreFailure) : Failure()
+    }
+}
+
+data class Redirects(val success: String, val error: String)
+
+interface SSOInitiateLoginUseCase {
+    suspend operator fun invoke(code: String, serverConfig: ServerConfig, redirects: Redirects? = null): SSOInitiateLoginResult
+}
+
+internal class SSOInitiateLoginUseCaseImpl(
+    private val ssoLoginRepository: SSOLoginRepository,
+    private val validateUUIDUseCase: ValidateUUIDUseCase
+) : SSOInitiateLoginUseCase {
+
+    override suspend fun invoke(code: String, serverConfig: ServerConfig, redirects: Redirects?): SSOInitiateLoginResult = suspending {
```

Comment:
I'm not really sure about the nullable Redirects, how about something like this
```kotlin
interface SSOInitiateLoginUseCase {
    sealed class Param {
        abstract val ssoCode: String
        abstract val serverConfig: ServerConfig
        data class NoRedirect(override val ssoCode: String, override val serverConfig: ServerConfig): Param()
        data class Redirect(override val ssoCode: String, val redirects: SSORedirects, override val serverConfig: ServerConfig): Param()
    }
    suspend operator fun invoke(param: Param): SSOInitiateLoginResult
}

internal class SSOInitiateLoginUseCaseImpl(
    private val ssoLoginRepository: SSOLoginRepository
) : SSOInitiateLoginUseCase {

    override suspend fun invoke(param: SSOInitiateLoginUseCase.Param): SSOInitiateLoginResult =
        with(param) {
            when (this) {
                is SSOInitiateLoginUseCase.Param.Redirect -> ssoLoginRepository.initiate(
                    ssoCode,
                    redirects.success,
                    redirects.error,
                    serverConfig
                )
                is SSOInitiateLoginUseCase.Param.NoRedirect -> ssoLoginRepository.initiate(ssoCode, serverConfig)
            }
        }.fold({
            if (it is NetworkFailure.ServerMiscommunication && it.kaliumException is KaliumException.InvalidRequestError) {
                if (it.kaliumException.errorResponse.code == HttpStatusCode.BadRequest.value)
                    SSOInitiateLoginResult.Failure.InvalidRedirect
                if (it.kaliumException.errorResponse.code == HttpStatusCode.NotFound.value)
                    SSOInitiateLoginResult.Failure.InvalidCode
            }
            SSOInitiateLoginResult.Failure.Generic(it)
        }, {
            SSOInitiateLoginResult.Success(it)
        })
}

```

Reply:
Sure, I was thinking about it but wasn't sure if it's not overcomplicating the code so I went to the shorter solution 😅 now I added your proposition to the code